Multiple Cloud Source Repo support

etanshaul opened the pull request:
**NOT yet ready for review.** Still pending cleanup/refactoring and testing; in the meantime you can see the screenshots below.

This PR add support in IJ for the GCP support of multiple cloud repositories. Any reference or reliance on a 'default' repository is also removed.

Upload to Cloud Repo dialog:
![image](https://cloud.githubusercontent.com/assets/1735744/21785859/4962d3e6-d68f-11e6-8f43-bffcba30d05c.png)

Checkout from Cloud Repo dialog:
![image](https://cloud.githubusercontent.com/assets/1735744/21785878/624dd3ce-d68f-11e6-8f74-975daae077ce.png)

Selector widget from the upload dialog ("create new repo" button is available):
![image](https://cloud.githubusercontent.com/assets/1735744/21786029/0e0c43e4-d690-11e6-8433-02f64ac3ad72.png)

Selector widget from the checkout dialog ("create new repo" button _not_ available):
![image](https://cloud.githubusercontent.com/assets/1735744/21786054/2935db12-d690-11e6-9899-7307c590e3b1.png)

Selector widget in loading state:
![image](https://cloud.githubusercontent.com/assets/1735744/21786193/c68d2c44-d690-11e6-8ee4-3f1c848499b9.png)

Selector widget in error state:
![image](https://cloud.githubusercontent.com/assets/1735744/21786222/e29637a0-d690-11e6-855d-3b4713d550b4.png)

Selector widget in empty state (no repos found):
![image](https://cloud.githubusercontent.com/assets/1735744/21786209/d39cc552-d690-11e6-9d78-96d6f74cb8e8.png)

etanshaul commented:
Thanks Rui for reviewing this. You can post feedback directly here on GitHub.

@patflynn - the 'create repo' button currently does not open a dialog. We have no API call for this yet - once the new Source API is release I plan up updating this. Instead, it opens up a browser window. Unfortunately the best I could do is open up the 'developer' page that lists all the repo's and has a "new repo" button at the top. This is because the web console's new repo button does not have a permanently linkable url (it just uses some javascript).

Regarding the spacing - I agree, I'll improve this.

etanshaul commented:
Rui, to install this build to check it out (but keep in mind things may change in the process of review) you can follow this steps:

1. Clone this repository
2. Switch to branch 'repository-selector'
3. Run IDEA with the plugin:
  `./gradlew clean buildPlugin runIdea`
4. Go to tools->Google Cloud Tools -> Upload to a Google Cloud Source Repository
    and
    VCS -> Checkout from Version Control -> Google Cloud

etanshaul commented:
@patflynn what are your thoughts on allowing a single GCP project to be uploaded to more than one Cloud Source Repo?

Context:

Previously, when there was only 1 'default' Cloud repo, IJ ran a check to see if there already existed a CSR remote. If the remote was there, then a dialog would come up:

<img width="427" alt="screen shot 2017-01-10 at 4 17 29 pm" src="https://cloud.githubusercontent.com/assets/1735744/21825083/4475dace-d750-11e6-8236-6cf759363da2.png">

Now that we support multiple CSR repos, we have 2 options:

1) keep the logic from before and only allow a given project to be associated with a single CSR remote
2) allow multiple remotes and update the check to occur after selection of the CSR repo - after selection it would check to see if the remote exists. We'd have to update the remote naming scheme - probably to make the remote name equal to the name of the repository ID.

Personally I keep going back and forth on these. On the one hand, since GCP now supports multiple repos, we could just allow a single project to be pushed to multiple repos. Other the other hand, when were talking about a single project, the same thing can (and usually is) achieved with branches.

patflynn commented:
What does the generic git integration for IJ allow?

On Tue, Jan 10, 2017 at 4:22 PM, Etan Shaul <notifications@github.com>
wrote:

> @patflynn <https://github.com/patflynn> what are your thoughts on
> allowing a single GCP project to be uploaded to more than one Cloud Source
> Repo?
>
> Context:
>
> Previously, when there was only 1 'default' Cloud repo, IJ ran a check to
> see if there already existed a CSR remote. If the remote was there, then a
> dialog would come up:
>
> [image: screen shot 2017-01-10 at 4 17 29 pm]
> <https://cloud.githubusercontent.com/assets/1735744/21825083/4475dace-d750-11e6-8236-6cf759363da2.png>
>
> Now that we support multiple CSR repos, we have 2 options:
>
>    1. keep the logic from before and only allow a given project to be
>    associated with a single CSR remote
>    2. allow multiple remotes and update the check to occur after
>    selection of the CSR repo - after selection it would check to see if the
>    remote exists. We'd have to update the remote naming scheme - probably to
>    make the remote name equal to the name of the repository ID.
>
> Personally I keep going back and forth on these. On the one hand, since
> GCP now supports multiple repos, we could just allow a single project to be
> pushed to multiple repos. Other the other hand, when were talking about a
> single project, the same thing can (and usually is) achieved with branches.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/google-cloud-intellij/pull/1184#issuecomment-271701891>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHf5HRus299i6zMFrfC5I4gZKLsSRlaYks5rQ_aUgaJpZM4Lezod>
> .
>

etanshaul commented:
If you have multiple remotes then the generic git integration allows you to select which remote to push to. I guess this answers my question.

Good point.

etanshaul commented:
@patflynn @joaoandremartins @aslo 

Other then the enhancement to allow uploading to multiple CSR repos (as discussed above), this is ready for initial review.

etanshaul commented:
I've added support for multiple CSR remotes: https://github.com/GoogleCloudPlatform/google-cloud-intellij/pull/1184/commits/b07b8a1324afbf11ae020d675b55d0407d97e951

Say I have a GCP project with the following repositories:

![image](https://cloud.githubusercontent.com/assets/1735744/21856750/198d2f6e-d7f1-11e6-85c2-3d874f05e213.png)

I can then upload my project to say 'test-repo-one', and locally, a git config with the following remote will be created:

![image](https://cloud.githubusercontent.com/assets/1735744/21856848/576a102c-d7f1-11e6-9528-131081f336c8.png)

I can then upload the same project to 'test-repo-2' as well, and 2 remotes will be present:
![image](https://cloud.githubusercontent.com/assets/1735744/21856885/7943b946-d7f1-11e6-8255-793addd5a617.png)

If I then try to upload the project again to 'test-repo-1' I'll get the following dialog:
![image](https://cloud.githubusercontent.com/assets/1735744/21856907/94b95cb2-d7f1-11e6-9bc7-c2606df02fe2.png)

Finally, if I manually update the remote url for 'test-repo-2' to some non GCP url to simulate a name collision and then try to upload to 'test-repo-2', I'll get the following dialog:
![image](https://cloud.githubusercontent.com/assets/1735744/21856970/cf216d7c-d7f1-11e6-9d01-17c3c78e096a.png)

hray2003 commented:
In "selector error state" and "selector empty state", what can user do to in order to move forward?

etanshaul commented:
@hray2003 - in both those cases, the user can either refresh, select a different project, or create a new repository. Perhaps the wording for the 'empty' state can be expanded to indicate that the user can create a new repository.

patflynn commented:
I don't fully understand the second case.

On Wed, Jan 11, 2017 at 2:38 PM, Etan Shaul <notifications@github.com>
wrote:

> @hray2003 <https://github.com/hray2003> - in both those cases, the user
> can either refresh, select a different project, or create a new repository.
> Perhaps the wording for the 'empty' state can be expanded to indicate that
> the user can create a new repository.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/google-cloud-intellij/pull/1184#issuecomment-271971905>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHf5HU6aC1744i4053rRIG3oa6siFMyWks5rRS-6gaJpZM4Lezod>
> .
>

etanshaul commented:
Which second case? The 'empty state'? This is just when a selected cloud project has no CSR repos associated with it.

joaoandremartins commented:
1) Can we rename "Repository Selector:" to just "Repository:"?

2) Maybe I'm missing something, but why are we prepending "cloud-platform-" to the remote name? To make sure it doesn't collide with any previously existing remote with the same name? To mark it as a GCR remote? (and then we'd probably run into problems if there were already other remotes named "cloud-platform-*", unlikely as that is)

3) Finally, I also don't understand the second case (where you manually modify the remote URL so it's not a GCR URL). What are you accomplishing there? Validating/enforcing that a GCR remote has a valid URL? Why does renaming the URL result in a name collision? Why are "upload to a GCR remote" and "add a GCR remote" related?

Also, the title of the window needs to be changed from "Google" to something else like "Google Cloud Source Repositories" and the text on the second window could also use some improvements, e.g., s/GCP remote/Google Cloud Source Repository remote. (and why are we telling we couldn't add a remote when the user is trying to push to an already added remote?)

joaoandremartins commented on google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties:
```diff
@@ -161,8 +161,9 @@ uploadtogcp.text=Upload to a Google Cloud Source Repository...
-uploadtogcp.selecttext=Select a Google Cloud Project
+uploadtogcp.alreadyexists=Project is already uploaded to the selected Google Cloud Source Repository ''{0}''.
+uploadtogcp.nongcp.remotename.collision=Failed to add GCP remote because of an existing non GCP remote name collision.\nTo fix, remove or rename the remote named ''{0}''.
+uploadtogcp.selecttext=Upload to Google Cloud Repositories
```
s/Google Cloud Repositories/Google Cloud Source Repository

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/ProjectRepositoriesModelItem.java:
```diff
@@ -0,0 +1,147 @@
+
+  private static final String CLOUD_SOURCE_API_ROOT_URL = "https://source.googleapis.com/";
+  private static final String CLOUD_SOURCE_API_LIST_URL = "v1/projects/{projectId}/repos";
+  static final String PANETHEON_CREATE_REPO_URL_PREFIX
```
Not being used anywhere else in this PR. Set to private?

etanshaul replied:
Its being used by the RepositorySelector class

joaoandremartins replied:
It's the only place it's being used at. Why not moving it over there?

etanshaul commented:
@joaoandremartins 

1. The "Repository Selector:" wording was a mistake and has since been updated (the screenshot is old). It now said "Cloud Repository".

2. You are right. I am prepending "cloud-platform-" to avoid naming collisions. If there happen to be collisions, then we show the following dialog:
![image](https://cloud.githubusercontent.com/assets/1735744/21864203/b3833bbe-d80e-11e6-89c0-649edf62beef.png)

3. This is just to similate the naming collision I just mentioned in (2). I specifically set it to a non GCP url because it if was a GCP url, then the user would see this dialog instead:
![image](https://cloud.githubusercontent.com/assets/1735744/21864231/d020165c-d80e-11e6-9d76-05d5c9acd7b8.png)
I'm not sure what you mean about "upload/add a GCR remote". Maybe lets talk more about this part offline.

I agree about your points about improving wording.

> and why are we telling we couldn't add a remote when the user is trying to push to an already added remote?

This is because our "Upload to GCP" functionality is _not_ for pushing to a GCP remote. Its for the initial project upload and for automatically creating the local git remote. Subsequently to push to the cloud repository the user is expected to use basic GIT functionality (either in IJ or otherwise).

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/ProjectRepositoriesModelItem.java:
```diff
@@ -0,0 +1,147 @@
+      Credential credential = user.getCredential();
+      HttpRequestInitializer initializer = httpRequest -> {
+        HttpHeaders headers = new HttpHeaders();
+        httpRequest.setConnectTimeout(5000);
```
Parameterize these timeout values in config.properties?

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/ProjectRepositoriesModelItem.java:
```diff
@@ -0,0 +1,147 @@
+      removeAllChildren();
+      List<Repo> repositories = response.getRepos();
+      if (!response.isEmpty() && repositories != null) {
+        for (Repo repo : repositories) {
```
Possibly looks better using Streams API?

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/RepositoryItem.java:
```diff
@@ -0,0 +1,35 @@
+/**
+ * UI for node in the repository selector representing a single GCP repository.
+ */
+public class RepositoryItem extends JBLabel {
```
We're only using this class once. Should we inline all these settings where we declare and initialize its instance (RepositorySelectorRenderer)?

etanshaul replied:
I think I will leave them in separate class as these can (and some of them are) be reused by multiple renders.

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/RepositorySelector.java:
```diff
@@ -0,0 +1,256 @@
+    this.user = user;
+    this.canCreateRepository = canCreateRepository;
+
+    getTextField().getEmptyText().setText("Enter a cloud repository ID...");
```
Send text to resource bundle?

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/RepositorySelector.java:
```diff
@@ -0,0 +1,256 @@
+      return null;
+    }
+
+    Enumeration repos = panel.getRepositories().children();
```
Is null check needed for getRepositories() return?

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/RepositorySelector.java:
```diff
@@ -0,0 +1,256 @@
+    }
+
+    Enumeration repos = panel.getRepositories().children();
+    while (repos.hasMoreElements()) {
```
If Streams API can be used with repos, it would greatly improve readability here.

etanshaul replied:
Sadly it can't since the API forces Enumerations.

joaoandremartins replied:
We could use Collections.list() to transform the enumeration into a collection. Would the extra pass be worth simpler code? I'm leaning to think yes, but I don't feel strongly about this.

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/RepositorySelector.java:
```diff
@@ -0,0 +1,256 @@
+
+  @Override
+  protected CustomizableComboBoxPopup getPopup() {
+    return this;
```
Do you want to return this or popup here?

etanshaul replied:
It needs a CustomizableComboBoxPopup, so "this" is correct.

joaoandremartins replied:
I wonder what is the use of a method that returns the reference of the object it's being called on?

etanshaul replied:
yeah its a strange remnant of legacy code. I'll look to see what can be done

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/RepositorySelector.java:
```diff
@@ -0,0 +1,256 @@
+    return SELECTOR_HEIGHT;
+  }
+
+  private class RepositoryPanel extends JPanel {
```
Wouldn't it be better to create a .form file for this?

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/RepositorySelectorRenderer.java:
```diff
@@ -0,0 +1,65 @@
+    if (value instanceof RepositoryModelItem) {
+      repositoryItem.setText(((RepositoryModelItem) value).getRepositoryId());
+      return repositoryItem;
+    } else if (value instanceof ResourceLoadingModelItem) {
```
Remove the elses for better readability?

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/ResourceLoadingModelItem.java:
```diff
@@ -16,12 +16,46 @@
 
 package com.google.cloud.tools.intellij.resources;
 
+import com.sun.istack.NotNull;
```
Wrong NotNull?

etanshaul replied:
nice catch

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/ResourceLoadingModelItem.java:
```diff
@@ -16,12 +16,46 @@
+/**
+ * This model item is shown when a resource manager call returns an empty result.
+ */
+class ResourceEmptyModelItem extends DefaultMutableTreeNode {
```
package-protected visibility on purpose?

joaoandremartins replied:
No custom behaviour seems to be tied to ResourceEmptyModelItem. Why not just use DefaultMutableTreeNode?

etanshaul replied:
It is adding a message String state to the class. I'll see if there's a way to refactor this.

joaoandremartins replied:
You could use new DefaultMutableTreeNode(Object object)

etanshaul replied:
yes

etanshaul replied:
The bigger point is that if you look at RepositorySelectorRenderer#getTreeCellRendererComponent its checking the type of DefaultMutableTreeNode so it can then decide which UX to show.

patflynn commented:
I was just talking with Etan about this offline and we're now thinking a
workflow might be more flexible with the following strategy:

- when 'uploading to GCP' always ask the user what they'd like to call the
remote we create.
 - For the first remote default it to 'origin'. (this is what the JB github
integration does)
 - For subsequent remotes possibly concatenate project name and repo name.
- don't try to map remote names to CSR repos.  Allow the user to create
multiple remotes pointing to the same uri if they want.



On Wed, Jan 11, 2017 at 3:04 PM, Etan Shaul <notifications@github.com>
wrote:

> @joaoandremartins <https://github.com/joaoandremartins>
>
>    1.
>
>    The "Repository Selector:" wording was a mistake and has since been
>    updated (the screenshot is old). It now said "Cloud Repository".
>    2.
>
>    You are right. I am prepending "cloud-platform-" to avoid naming
>    collisions. If there happen to be collisions, then we show the following
>    dialog:
>    [image: image]
>    <https://cloud.githubusercontent.com/assets/1735744/21864203/b3833bbe-d80e-11e6-89c0-649edf62beef.png>
>    3.
>
>    This is just to similate the naming collision I just mentioned in (2).
>    I specifically set it to a non GCP url because it if was a GCP url, then
>    the user would see this dialog instead:
>    [image: image]
>    <https://cloud.githubusercontent.com/assets/1735744/21864231/d020165c-d80e-11e6-9d76-05d5c9acd7b8.png>
>    I'm not sure what you mean about "upload/add a GCR remote". Maybe lets
>    talk more about this part offline.
>
> I agree about your points about improving wording.
>
> and why are we telling we couldn't add a remote when the user is trying to
> push to an already added remote?
>
> This is because our "Upload to GCP" functionality is *not* for pushing to
> a GCP remote. Its for the initial project upload and for automatically
> creating the local git remote. Subsequently to push to the cloud repository
> the user is expected to use basic GIT functionality (either in IJ or
> otherwise).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/google-cloud-intellij/pull/1184#issuecomment-271978616>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHf5HR3Lw4qi7L6t0WyZysQtN0EkERpsks5rRTXOgaJpZM4Lezod>
> .
>

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/ResourceLoadingModelItem.java:
```diff
@@ -16,12 +16,46 @@
+ * This model item represents a node shown in when an error occurs attempting to query a resource
+ * manager.
+ */
+class ResourceErrorModelItem extends DefaultMutableTreeNode {
```
Ditto as above.

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/ResourceSelectorEmptyItem.java:
```diff
@@ -0,0 +1,39 @@
+/**
+ * UI for the node that displays information if a resource manager call returns an empty result.
+ */
+public class ResourceSelectorEmptyItem extends JBLabel {
```
Also only used once in this PR. Inline?

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/ResourceSelectorErrorItem.java:
```diff
@@ -29,9 +29,9 @@
  * UI for the node that displays error information if a resource manager call fails.
  */
-class ProjectSelectorErrorItem extends JBLabel {
+public class ResourceSelectorErrorItem extends JBLabel {
```
Inline?

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/CloneGcpDialog.form:
```diff
@@ -51,20 +51,34 @@
+          <grid row="1" column="0" row-span="1" col-span="1" vsize-policy="0" hsize-policy="0" anchor="8" fill="0" indent="0" use-parent-layout="false"/>
+        </constraints>
+        <properties>
+          <text value="Cloud Repository:"/>
```
Rename "Cloud Repository" and "Cloud Project" to "Repository" and "Cloud" in this menu?

etanshaul replied:
I like the "Cloud" prefix here to disambiguate between the gcp project/repo and the local project/repo. Without it, it may be confusing.

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/CloneGcpDialog.java:
```diff
@@ -106,21 +108,21 @@ private void initComponents() {
-          return super.getInitialFile();
-        }
-      }
+                                          .BrowseFolderActionListener<JTextField>(
```
Spacing doesn't seem right according to style guide. Did google-java-format do this?

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/CloneGcpDialog.java:
```diff
@@ -106,21 +108,21 @@ private void initComponents() {
-      }
+                                          .BrowseFolderActionListener<JTextField>(
+                                          fcd.getTitle(), fcd.getDescription(), parentDirectory,
+                                          project, fcd, TextComponentAccessor.TEXT_FIELD_WHOLE_TEXT) {
```
Longer than 100 lines.

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/CloneGcpDialog.java:
```diff
@@ -150,7 +152,22 @@ protected String getHelpId() {
+      setErrorText("Invalid Cloud Project selected.");
+      setOKActionEnabled(false);
+      return;
+    } else if (!StringUtil.isEmpty(repositorySelector.getText())
```
Remove elses for readability?

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/CloneGcpDialog.java:
```diff
@@ -150,7 +152,22 @@ protected String getHelpId() {
+      setErrorText(null);
+      setOKActionEnabled(false);
+      return;
+    } else if (parentDirectory.getText().length() == 0 || directoryName.getText().length() == 0) {
```
Could parentDirectory().getText() ever return null?

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/CloneGcpDialog.java:
```diff
@@ -167,34 +184,37 @@ private void updateButtons() {
+    projectSelector = new ProjectSelector();
+    projectSelector.setMinimumSize(new Dimension(300, 0));
+    projectSelector.getDocument().addDocumentListener(new DocumentAdapter() {
       @SuppressWarnings("ConstantConditions") // This suppresses an invalid nullref warning for
```
Is this still needed?

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/CloneGcpDialog.java:
```diff
@@ -204,15 +224,29 @@ protected void textChanged(DocumentEvent event) {
+
+    repositorySelector.getDocument().addDocumentListener(new DocumentAdapter() {
+      @Override
+      protected void textChanged(DocumentEvent e) {
```
s/e/event?

aslo commented on google-cloud-tools-plugin/build.gradle:
```diff
@@ -54,6 +54,7 @@ dependencies {
         exclude group: 'com.google.guava', module: 'guava'
     }
     compile files('lib/google-api-services-source.jar')
+//    compile files('lib/libsource-v1-java.jar')
```
Is the idea that this line will be uncommented when the CSR API is ready?

etanshaul replied:
thats not actually the correct API either. but yeah - this will be updated to use the correct API before merging.

joaoandremartins replied:
Are you planning on removing this?

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/UploadSourceAction.java:
```diff
@@ -136,31 +140,34 @@ private static void uploadProjectToGoogleCloud(@NotNull final Project project,
+        ? findRemote(gitRepository.getRemotes(), dialog.getRepositoryId())
+        : Optional.empty();
+    if (remote.isPresent() && !GcpHttpAuthDataProvider.hasGcpUrl(remote.get().getUrls())) {
+      Messages.showErrorDialog(project,
```
Bundleize "Google".

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/UploadSourceAction.java:
```diff
@@ -136,31 +140,34 @@ private static void uploadProjectToGoogleCloud(@NotNull final Project project,
       return;
     }
 
+    Optional<GitRemote> remote = gitDetected
```
If !gitDetected, can you return right away instead of having this assignment?

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/UploadSourceAction.java:
```diff
@@ -136,31 +140,34 @@ private static void uploadProjectToGoogleCloud(@NotNull final Project project,
+    Optional<GitRemote> remote = gitDetected
+        ? findRemote(gitRepository.getRemotes(), dialog.getRepositoryId())
+        : Optional.empty();
+    if (remote.isPresent() && !GcpHttpAuthDataProvider.hasGcpUrl(remote.get().getUrls())) {
```
Have a single top level if (remote.isPresent()) ?

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/UploadSourceAction.java:
```diff
@@ -176,20 +183,17 @@ public void run(@NotNull ProgressIndicator indicator) {
-          });
+          SwingUtilities.invokeLater(() ->
+              Messages.showErrorDialog(project,
+                  GctBundle.message("uploadtogcp.failedtocreategit"), "Google"));
```
Bundleize Google and add something else, i.e., Google Cloud Source Repositories?

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/UploadSourceDialog.java:
```diff
@@ -73,18 +82,57 @@ public CredentialedUser getCredentialedUser() {
+
+    repositorySelector.getDocument().addDocumentListener(new DocumentAdapter() {
+      @Override
+      protected void textChanged(DocumentEvent e) {
```
s/e/event?

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/UploadSourceDialog.java:
```diff
@@ -73,18 +82,57 @@ public CredentialedUser getCredentialedUser() {
+//    repositorySelector.setEnabled(false);
+  }
+
+  private void updateButtons() {
```
Remove elses?

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/UploadSourceDialog.java:
```diff
@@ -73,18 +82,57 @@ public CredentialedUser getCredentialedUser() {
+  private void updateButtons() {
+    if (!StringUtil.isEmpty(projectSelector.getText())
+        && projectSelector.getSelectedUser() == null) {
+      setErrorText("Invalid Cloud Project selected.");
```
Bundleize text?

aslo commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/ProjectRepositoriesModelItem.java:
```diff
@@ -0,0 +1,147 @@
+
+  public void loadRepositories(boolean empty) {
+
+    // TODO remove
```
Don't forget to RM

etanshaul replied:
Yea I should have mentioned prior to asking for review - I purposefully left in a couple of TODO's (will remove soon) and I haven't bundleized any messaged yet.

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/UploadSourceDialog.java:
```diff
@@ -73,18 +82,57 @@ public CredentialedUser getCredentialedUser() {
+      setErrorText(null);
+      setOKActionEnabled(false);
+      return;
+    }
```
Is this the same updateButtons() as CloneGcpDialog? Can anything be done about that?

etanshaul replied:
Part of it is the same. I think I can collapse the validation into the selector classes so the code can be reused.

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/CloneGcpDialog.java:
```diff
@@ -52,22 +54,22 @@
 public class CloneGcpDialog extends DialogWrapper {
+
   private static final Logger LOG = Logger.getInstance(CloneGcpDialog.class);
   public static final String INVALID_FILENAME_CHARS = "[/\\\\?%*:|\"<>]";
```
This variable doesn't seem to be used anywhere else other than this class.
LOG is also showing as unused

etanshaul commented:
@patflynn I've updated the experience to reflect our discussion:

There is now always a "remote name" selector:
![image](https://cloud.githubusercontent.com/assets/1735744/22031228/0dc0bbda-dcae-11e6-91c4-b7d60637e24c.png)

The remote is auto-populated with a suggestion:
- "origin" if there is no remote named "origin" already
- "google-cloud-platform-{repository_name}" if there is already a remote named "origin"

If the user forces a conflict (by manually typing in a remote name that already exists) the following will be shown:
![image](https://cloud.githubusercontent.com/assets/1735744/22031447/a5c6ee4a-dcae-11e6-9120-4b6d144b9e0e.png)

etanshaul commented:
@patflynn @joaoandremartins @aslo 

What are your thoughts on the wording of this dialog, specifically the "Remote name" section:
![image](https://cloud.githubusercontent.com/assets/1735744/22039820/cae46b92-dcce-11e6-8b21-e258931241f8.png)

I was going through the process of testing out the flow as a normal user and realized it may not be totally clear that this dialog will generate a new remote for the user. I'm concerned that a user may confuse this with us requiring an existing remote to exist and that its up to the user to type in the matching name.

Here are a few options:

- Change the label from "Remote name" to "New remote name"
- Have placeholder text that says something like "enter a remote name to be created for this repository"
- Have some labeling for the dialog explaining that this creates a new git repo (if one doesn't exist), and a new remote

joaoandremartins commented:
Does it have to be a new remote? If it doesn't, then I'd be OK with not extra wording -- if the remote doesn't exist yet, we create it for the user.

etanshaul commented:
Yes, it does have to be a new remote. Otherwise its an error.

joaoandremartins commented:
I think "New remote name" looks good. The only problem is if the user doesn't understand why it needs to be a new repo (e.g., if they haven't checked out the remote repo but want to push there right away. not sure how often/if this would happen at all?). Then we could need some text.

etanshaul commented:
> e.g., if they haven't checked out the remote repo but want to push there right away. not sure how often/if this would happen at all?

I'd like to understand this scenario. Not having checked out the remote repo is the majority of use cases. This feature is precisely for "intializing" the remote repo by pushing the project to it and establishing the local remote (sounds like an oxymoron).

joaoandremartins commented:
I think the case you're referring to is when the user doesn't _already_ have a remote repo with that code, and is starting fresh.

If the code is already in the remote, but the remote somehow isn't synced locally (does this happen at all?), the user might want to use that "Upload to GCSR" dialog to upload it there. Though, thinking about it, I don't think this happens frequently?

Also, what's the story when you check out the remote with git or gcloud? Will we still require the users to create a new remote?

etanshaul commented:
> If the code is already in the remote, but the remote somehow isn't synced locally (does this happen at all?), the user might want to use that "Upload to GCSR" dialog to upload it there. Though, thinking about it, I don't think this happens frequently?

This dialog will always create a new remote. Whether or not we allow uploading to a cloud repo that already has code is up for debate: previously it would _not_ allow this. After talking to Patrick we decided that it would (a new remote would still be created pointing to the same cloud url). However, I am going back on this now because there are cases where this could fail: e.g. non-fast-foward changes, which is tough to detect before hand.

> Also, what's the story when you check out the remote with git or gcloud? Will we still require the users to create a new remote?

Yes, if they use this dialog in this case (and they shouldn't really) a new remote will be forced.

If I had to summarize what this dialog does it would be as follows:

The dialog creates a new local git repository if one does not yet exist, creates a new remote pointing to the Google Cloud Repository, and finally pushes the code to this remote.

joaoandremartins commented:
I wonder if it would make sense to add a new issue to iron that bit out, i.e., if the user already has a remote with the code that isn't mirrored in a way our IJ plugin can't use it out of the bat?

In any case, since that dialog will always create a new remote and users using this feature who don't have a correctly set up remote, I think we can do without the extra text? Patrick or Rui might have a different take on that, though.

patflynn commented:
I'm kind of OK with either.

Could hover text or some extra wording above the fields clarify things?

On Tue, Jan 17, 2017 at 4:54 PM, João Martins <notifications@github.com>
wrote:

> I wonder if it would make sense to add a new issue to iron that bit out,
> i.e., if the user already has a remote with the code that isn't mirrored in
> a way our IJ plugin can't use it out of the bat?
>
> In any case, since that dialog will always create a new remote and users
> using this feature who don't have a correctly set up remote, I think we can
> do without the extra text? Patrick or Rui might have a different take on
> that, though.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/google-cloud-intellij/pull/1184#issuecomment-273312233>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHf5Hfn8q1fxrEetfVyh_DRxOb4rSUcdks5rTTiJgaJpZM4Lezod>
> .
>

etanshaul commented:
Here is what I came up with, with help from @joaoandremartins:

Updated action text (to make it clear that this is for "Setup" and not for normal "uploading / pushing"):
![image](https://cloud.githubusercontent.com/assets/1735744/22042336/9ebddbd4-dcd8-11e6-948d-3ec5c2a517f0.png)

Updated remote name label to make it clear that its creating a new one:
![image](https://cloud.githubusercontent.com/assets/1735744/22042342/a95caf66-dcd8-11e6-8366-9bcdb354864d.png)

etanshaul commented:
I've updated the error messages to be notification popups (instead of an error dialog) and to show the contents of the git error output so the user can debug the problem.

Here is an example of the user attempting to use the dialog to a project to an existing git repo:

![image](https://cloud.githubusercontent.com/assets/1735744/22075161/de03648a-dd78-11e6-862f-48314918b927.png)

And after expanding, you can see the full message in a scrollable pane:
![image](https://cloud.githubusercontent.com/assets/1735744/22075188/fb58357e-dd78-11e6-8378-f778b0103017.png)

etanshaul commented:
@patflynn @joaoandremartins @aslo 

This is ready for review. All functionality has been implemented. Only outstanding task is test coverage.

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/RepositorySelector.java:
```diff
@@ -0,0 +1,292 @@
+
+  @Nullable
+  public String getSelectedRepository() {
+    if (StringUtil.isEmpty(getText())) {
```
What does this condition assure? Seems like we could do without it, unless the callers treat nulls and empty strings differently. Or if the assumption is being made that if the text is empty, getRepositories()/children() returns null or an empty collection, in which case, the condition should probably target those returns.

etanshaul replied:
If getText() is empty, this means that the user hasn't selected a repo from the tree, making all the subsequent checks unnecessary.

etanshaul replied:
if the text is empty it doesn't mean that the repository tree is empty, it just means that the user hasn't made a selection from the tree so checking each node is pointless

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/RepositorySelector.java:
```diff
@@ -0,0 +1,292 @@
+      return null;
+    }
+
+    Enumeration repos = ((RepositoryPanel) panel).getRepositories().children();
```
Is a null check for getRepositories() needed?

etanshaul replied:
no, added a `@NotNull` to make it explicit

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/RepositorySelector.java:
```diff
@@ -0,0 +1,292 @@
+    private ProjectRepositoriesModelItem repositories;
+
+    RepositoryPanel(String cloudProject) {
+      repositories
```
Is the new line needed here?

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/RepositorySelector.java:
```diff
@@ -0,0 +1,292 @@
+    RepositoryPanel(String cloudProject) {
+      repositories
+          = new ProjectRepositoriesModelItem(cloudProject, user);
+      projectRootNode = new DefaultMutableTreeNode("root");
```
Is "root" a special tree node?

etanshaul replied:
no, the string "root" has no significance, its there just for clarity. I think the tree will behave differently if I were to pass in null.

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/RepositorySelector.java:
```diff
@@ -0,0 +1,292 @@
+
+      RepositorySelectorRenderer renderer = new RepositorySelectorRenderer();
+      renderer.setOpenIcon(GoogleCloudToolsIcons.CLOUD);
+      renderer.setClosedIcon(GoogleCloudToolsIcons.CLOUD);
```
Do we want to have a different icon here? What does the closed icon represent anyway?

etanshaul replied:
yea, its for the icon that is used when the tree is expanded vs collapsed. For whatever reason theres no way to set both at once.

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/RepositorySelector.java:
```diff
@@ -0,0 +1,292 @@
+        if (node != null && node instanceof RepositoryModelItem) {
+          RepositoryModelItem repoNode = (RepositoryModelItem) node;
+          RepositorySelector.this.setText(repoNode.getRepositoryId());
+          ApplicationManager.getApplication().invokeLater(RepositorySelector.this::hidePopup);
```
wouldn't this::hidePopup suffice?

etanshaul replied:
no, we are in an inner class here, so I need a reference to the enclosing class

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/RepositorySelector.java:
```diff
@@ -0,0 +1,292 @@
+      buttonPanel.setLayout(new BoxLayout(buttonPanel, BoxLayout.LINE_AXIS));
+      buttonPanel.setBorder(BorderFactory.createEmptyBorder(5, 5, 5, 5));
+
+      if (canCreateRepository) {
```
If you can't create a repository, what happens?

etanshaul replied:
then it just omits the "create repository" button. this is as expected - there are 2 dialogs where this widget is used, 1 where we need the button, and 1 where we don't.

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/RepositorySelector.java:
```diff
@@ -0,0 +1,292 @@
+    }
+
+    private void setLoader() {
+      ApplicationManager.getApplication().invokeAndWait(() -> {
```
Why does this need to be done in a separate thread?

etanshaul replied:
because it modifies the UI

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/RepositorySelectorRenderer.java:
```diff
@@ -0,0 +1,71 @@
+  @Override
+  public Component getTreeCellRendererComponent(JTree tree, Object value, boolean selected,
+      boolean expanded, boolean leaf, int row, boolean hasFocus) {
+    if (value instanceof RepositoryModelItem) {
```
This still strikes me as a bit weird style of programming. Is there no better alternative to these class checks? And is there a reason we can't simply initialize the *Item's here instead of in the constructor?

etanshaul replied:
yea its super weird, but its because of how the renderers work:

`add`ing to a tree requires you to pass a TreeNode. But using a custom renderer then allows you to "override" this with a custom component for the node (like a JLabel or JPanel). This overriding happen in the renderer - its caught by getTreeCellRendererComponent

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/ResourceLoadingModelItem.java:
```diff
@@ -16,12 +16,46 @@
+ * This model item represents a node shown when an error occurs attempting to query a resource
+ * manager.
+ */
+class ResourceErrorModelItem extends DefaultMutableTreeNode {
```
nit: Create new file for this class?

etanshaul replied:
will leave as is since all these classes are very trivial so I'd prefer to just group them all in the same file (there are 3 now)

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/CloneGcpDialog.java:
```diff
@@ -150,51 +150,77 @@ protected String getHelpId() {
       return;
     }
+
     File file = new File(parentDirectory.getText(), directoryName.getText());
```
Would it make sense to refactor here to use java.nio? It would come with the annoyance of having to catch InvalidPathExceptions from paths with special characters on Windows...

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/CloneGcpDialog.java:
```diff
@@ -150,51 +150,77 @@ protected String getHelpId() {
     }
 
-    return GcpHttpAuthDataProvider.getGcpUrl(repositoryUrl.getText());
+    // TODO what if the repo is null??
```
Do you want to address this TODO in this PR?

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/GcpCheckoutProvider.java:
```diff
@@ -183,6 +183,7 @@ private static boolean doClone(@NotNull Project project, @NotNull ProgressIndica
       @NotNull String sourceRepositoryUrl) {
     indicator.setIndeterminate(false);
     GitLineHandlerListener progressListener = GitStandardProgressAnalyzer.createListener(indicator);
+
```
heh nice

etanshaul replied:
lol...sometimes you just need some space

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/GcpHttpAuthDataProvider.java:
```diff
@@ -113,41 +111,18 @@ public void run() {
-    }
-    return null;
+  public static String getGcpUrl(String projectId, String repositoryId) {
+    return "https://source.developers.google.com/p/" + projectId + "/r/" + repositoryId + "/";
```
Parameterize this URL somewhere?

joaoandremartins replied:
What do you think about this?

etanshaul replied:
I'll extract the prefix to member variable

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/UploadSourceDialog.java:
```diff
@@ -0,0 +1,187 @@
+ * Shows the upload to Google Cloud Source Repositories dialog for initializing a project on
+ * a GCP repository.
+ */
+public class UploadSourceDialog extends DialogWrapper {
```
Rename to SetupGCSRepositoryDialog?

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/UploadSourceDialog.java:
```diff
@@ -0,0 +1,187 @@
+
+    repositorySelector.getDocument().addDocumentListener(new DocumentAdapter() {
+      @Override
+      protected void textChanged(DocumentEvent e) {
```
s/e/event

etanshaul commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/RepositoryRemotePanel.java:
```diff
@@ -0,0 +1,88 @@
+    }
+  }
+
+  private boolean hasOriginRemote() {
```
since there wont be any performance improvement (not that its a concern even), I'll leave it since I like tiny methods.

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/ProjectRepositoriesModelItem.java:
```diff
@@ -0,0 +1,78 @@
+ * project.
+ */
+public class ProjectRepositoriesModelItem extends DefaultMutableTreeNode {
+
```
Nit: extra space?

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/RepositorySelector.java:
```diff
@@ -0,0 +1,301 @@
+  }
+
+  @Nullable
+  public String getSelectedRepository() {
```
How about we change the return type to Optional<> here, to avoid returning null?

etanshaul replied:
In theory this is good. But we are still using nullables all over the place. Also if the API itself called from `getSelectedRepository` used Optionals I wouldn't hesitate to just return them, but I don't see much value in wrapping in an Optional now.

joaoandremartins replied:
If we never stop using nullables, then we're never going to stop using them. :)
We control the API that's calling this method, so we can adapt it to use Optional.
I feel strongly we should start replacing every null return to Optional wherever we see them.

etanshaul replied:
I started going down this path but it led down a rabbit hole of changing too much code in this PR (the consumers of this method are also consuming other code that uses nullables and I would want to change it all simultaneously so we don't have coding mixing and matching).

I'd rather not make this PR even bigger. Thoughts?

joaoandremartins replied:
Not sure if adding a TODO is worth in this case. I think it's a fact that at some point, we'll go over all these instances and make them Java 8 friendly, so it should be OK.

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/SetupCloudRepositoryDialog.java:
```diff
@@ -0,0 +1,187 @@
+      @NotNull String title, @NotNull String okText) {
+    super(project, true);
+
+    this.gitRepository = gitReository;
```
s/gitReository/gitRepository?

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/SetupCloudRepositoryDialog.java:
```diff
@@ -0,0 +1,187 @@
+  }
+
+  /**
+   * Return the credentialeduser that owns the ID returned from {@link #getProjectId()}.
```
nit add a space to credentialeduser?

etanshaul commented:
@joaoandremartins see https://github.com/GoogleCloudPlatform/google-cloud-intellij/pull/1184/commits/d6b25811ee2549176acefefa47f1ff02de68b1a6

I made some non-trivial changes to how we are executing asynchronous code. This is mainly for testability (there was no good way to test the asynchronous code previously without injecting a Thread.sleep into the unit test; now the asynchronous part is easily mocked away), but also for better design (hopefully) and more idiomatic java 8.

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/ProjectRepositoriesModelItem.java:
```diff
@@ -53,26 +50,36 @@ public ProjectRepositoriesModelItem(@NotNull String cloudProject,
+    cloudRepositoryService
+        .listAsync(user, cloudProject)
+        .thenAccept(response -> {
+          if (response == null) {
```
Can response ever be null?

etanshaul replied:
good question. I think we can remove this, but I'm checking.

etanshaul replied:
removed the null check. it will either return a response or throw an exception, never null.

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/resources/ProjectRepositoriesModelItem.java:
```diff
@@ -53,26 +50,36 @@ public ProjectRepositoriesModelItem(@NotNull String cloudProject,
+          add(new ResourceErrorModelItem(GctBundle.message("cloud.repository.list.error")));
+
+          onComplete.run();
+          return null;
```
Is this the method return? If so, should we make it return; only?

etanshaul replied:
it requires a return value. I couldn't find a cleaner way to handle the exception case. The exceptionally method requires a lambda returning another CompletableFuture since its a fluent API.

joaoandremartins replied:
Thanks for the explanation. Then I'd return Optional.empty().

etanshaul replied:
I can't return that. The return value is a CompletableFuture so it needs to be of that type (or null).

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/CloudRepositoryService.java:
```diff
@@ -45,29 +46,38 @@
+        credential.initialize(httpRequest);
+      };
+
+      HttpTransport httpTransport = GoogleNetHttpTransport.newTrustedTransport();
```
The API makes us go this low? I wonder if the new one takes care of this for us..

etanshaul replied:
hopefully

joaoandremartins commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/vcs/CloudRepositoryService.java:
```diff
@@ -88,3 +98,6 @@ public ListReposResponse list(CredentialedUser user, String cloudProject)
   }
 }
+
+class CloudRepositoryServiceException extends RuntimeException {
```
Why did you create this new exception?
